File: reference/rn-engine/src/games/sudoku/gameplay/analysis/techniqueModules/xyzWing.ts
import { popcount } from '../bitmask';
import { arePeerIndexes, buildCandidateEliminationMove, getCellCandidatesByIndex, peerIntersectionIndexes } from '../techniqueHelpers';
import type { SudokuTechniqueDispatcher } from '../techniqueModuleTypes';

export const xyzWingTechnique: SudokuTechniqueDispatcher = {
  technique: 'xyz-wing',
  tier: 'expert',
  findMove(state) {
    const pivotCells = state.board
      .map((value, index) => ({ value, index }))
      .filter(({ value, index }) => value === 0 && popcount[state.candidateMask[index]] === 3)
      .map(({ index }) => ({ index, candidates: getCellCandidatesByIndex(state, index) }));
    const wings = state.board
      .map((value, index) => ({ value, index }))
      .filter(({ value, index }) => value === 0 && popcount[state.candidateMask[index]] === 2)
      .map(({ index }) => ({ index, candidates: getCellCandidatesByIndex(state, index) }));

    const complexity = pivotCells.length;

    for (const pivot of pivotCells) {
      const pivotPeers = wings.filter((wing) => arePeerIndexes(wing.index, pivot.index));
      for (let leftIndex = 0; leftIndex < pivotPeers.length; leftIndex += 1) {
        for (let rightIndex = leftIndex + 1; rightIndex < pivotPeers.length; rightIndex += 1) {
          const leftWing = pivotPeers[leftIndex];
          const rightWing = pivotPeers[rightIndex];
          const leftSubset = leftWing.candidates.every((digit) => pivot.candidates.includes(digit));
          const rightSubset = rightWing.candidates.every((digit) => pivot.candidates.includes(digit));
          if (!leftSubset || !rightSubset) {
            continue;
          }

          const wingUnion = Array.from(new Set([...leftWing.candidates, ...rightWing.candidates])).sort((a, b) => a - b);
          if (wingUnion.length !== 3 || wingUnion.some((digit) => !pivot.candidates.includes(digit))) {
            continue;
          }

          const commonDigits = leftWing.candidates.filter((digit) => rightWing.candidates.includes(digit));
          if (commonDigits.length !== 1) {
            continue;
          }

          const zDigit = commonDigits[0];
          const move = buildCandidateEliminationMove({
            technique: 'xyz-wing',
            eliminations: peerIntersectionIndexes([pivot.index, leftWing.index, rightWing.index])
              .filter((index) => ![pivot.index, leftWing.index, rightWing.index].includes(index))
              .filter((index) => getCellCandidatesByIndex(state, index).includes(zDigit))
              .map((index) => ({ index, digit: zDigit })),
            evidenceCells: [pivot.index, leftWing.index, rightWing.index],
            complexity,
          });
          if (move) {
            return move;
          }
        }
      }
    }

    return null;
  },
};

File: engine/src/main/kotlin/com/quietgrid/engine/sudoku/techniques/XyzWing.kt
package com.quietgrid.engine.sudoku.techniques

import com.quietgrid.engine.core.Difficulty
import com.quietgrid.engine.sudoku.*

object XyzWingTechnique : SudokuTechniqueDispatcher {
    override val technique = SudokuTechnique.XYZ_WING
    override val tier = Difficulty.EXPERT

    private data class Cell(val index: Int, val candidates: List<Int>)

    override fun findMove(state: SudokuBitmaskState): SudokuCanonicalMove? {
        val pivotCells = state.board.indices.filter { state.board[it] == 0 && popcount[state.candidateMask[it]] == 3 }
            .map { Cell(it, getCellCandidatesByIndex(state, it)) }
        val wings = state.board.indices.filter { state.board[it] == 0 && popcount[state.candidateMask[it]] == 2 }
            .map { Cell(it, getCellCandidatesByIndex(state, it)) }
        val complexity = pivotCells.size

        for (pivot in pivotCells) {
            val pivotPeers = wings.filter { arePeerIndexes(it.index, pivot.index) }
            for (leftIndex in pivotPeers.indices) {
                for (rightIndex in (leftIndex + 1) until pivotPeers.size) {
                    val leftWing = pivotPeers[leftIndex]
                    val rightWing = pivotPeers[rightIndex]
                    if (!leftWing.candidates.all { pivot.candidates.contains(it) } || !rightWing.candidates.all { pivot.candidates.contains(it) }) continue

                    val wingUnion = (leftWing.candidates + rightWing.candidates).toSortedSet().toList()
                    if (wingUnion.size != 3 || wingUnion.any { !pivot.candidates.contains(it) }) continue

                    val commonDigits = leftWing.candidates.filter { rightWing.candidates.contains(it) }
                    if (commonDigits.size != 1) continue
                    val zDigit = commonDigits[0]

                    val eliminations = peerIntersectionIndexes(listOf(pivot.index, leftWing.index, rightWing.index))
                        .filter { it != pivot.index && it != leftWing.index && it != rightWing.index }
                        .filter { getCellCandidatesByIndex(state, it).contains(zDigit) }
                        .map { it to zDigit }

                    val move = buildCandidateEliminationMove(
                        SudokuTechnique.XYZ_WING, eliminations,
                        listOf(pivot.index, leftWing.index, rightWing.index), null, complexity,
                    )
                    if (move != null) return move
                }
            }
        }
        return null
    }
}
